src/network/clientpackethandler.cpp

```diff
@@ -217,9 +217,30 @@ void Client::handleCommand_AccessDenied(NetworkPacket* pkt)
 		*pkt >> denyCode;
 		if (denyCode == SERVER_ACCESSDENIED_CUSTOM_STRING) {
 			*pkt >> m_access_denied_reason;
-		}
-		else if (denyCode < SERVER_ACCESSDENIED_MAX) {
-			m_access_denied_reason = accessDeniedStrings[denyCode];
+		} else {
+			if (denyCode < SERVER_ACCESSDENIED_MAX) {
+				m_access_denied_reason = accessDeniedStrings[denyCode];
+				if (denyCode == SERVER_ACCESSDENIED_SHUTDOWN ||
+						denyCode == SERVER_ACCESSDENIED_CRASH) {
+					*pkt >> m_access_denied_reason;
+				} else {
+					std::string reason_str;
+					*pkt >> reason_str;
+				}
+
+				u8 reconnect = 0;
+				*pkt >> reconnect;
+				m_access_denied_reconnect = reconnect & 1;
+			} else {
+				// Allow us to add new error messages to the
+				// protocol without raising the protocol version, if we want to.
+				// Until then (which may be never), this is outside
+				// of the defined protocol.
+				*pkt >> m_access_denied_reason;
+				if (m_access_denied_reason == "") {
+					m_access_denied_reason = "Unknown";
+				}
+			}
 		}
```

Comment:
Globally the handling shout be rewritten like this to be better:
if (denyCode == SERVER_ACCESSDENIED_CUSTOM_STRING) { extract the string }
else if (denyCode == SERVER_ACCESSDENIED_SHUTDOWN || denyCode == SERVER_ACCESSDENIED_CRASH) { extract string & reconnect }
else if (denyCode >= SERVER_ACCESSDENIED_MAX) { do unknown trigger }